lib/puppet/util/windows/eventlog.rb

```diff
@@ -0,0 +1,180 @@
+require 'ffi'
+
+# Puppet::Util::Windows::EventLog needs to be requirable without having loaded
+# any other parts of Puppet so it can be leveraged independently by the code
+# that runs Puppet as a service on Windows.
+#
+# For this reason we:
+# - Define Puppet::Util::Windows
+# - Replicate logic that exists elsewhere in puppet/util/windows
+# - Raise generic RuntimeError instead of Puppet::Util::Windows::Error if its not defined
+module Puppet; module Util; module Windows ; end ; end ; end
+
+class Puppet::Util::Windows::EventLog
+  extend FFI::Library
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa363679(v=vs.85).aspx
+  EVENTLOG_ERROR_TYPE       = 0x0001
+  EVENTLOG_WARNING_TYPE     = 0x0002
+  EVENTLOG_INFORMATION_TYPE = 0x0004
+
+  # Register an event log handle for the application
+  # @param source_name [String] the name of the event source to retrieve a handle for
+  # @return [void]
+  # @api public
+  def initialize(source_name = 'Puppet')
+    @eventlog_handle = RegisterEventSourceW(nil, wide_string(source_name))
+    if @eventlog_handle == NULL_HANDLE
+      raise EventlogError.new("failed to open Windows eventlog")
+    end
+  end
+
+  # Close this instance's event log handle
+  # @return [void]
+  # @api public
+  def close
+    result_of_close = DeregisterEventSource(@eventlog_handle)
+    if result_of_close == WIN32_FALSE
+      raise EventlogError.new("failed to close Windows eventlog")
```

Comment:
I believe we call `close` here from the `:eventlog` destination type's `close` method [defined here](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/log/destinations.rb#L249-L254). Each destination type optionally defines `close`. This is in turn called by [`Puppet::Util::Log::close`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/log.rb#L52-L60), used by various methods in that class.

I agree we probably always want to set the handle to nil - given what you say about minimal real-world consequences I'm inclined to remove the `raise` altogether and instead do something like
```ruby
begin
  DeregisterEventSource(@eventlog_handle)
ensure
  @eventlog_handle = nil
end
```
thoughts?